stable/agent/templates/horizontal-pod-autoscaler.yaml

```diff
@@ -0,0 +1,42 @@
+{{- if and (.Values.horizontalPodAutoscaler.enabled) (.Values.horizontalPodAutoscaler.types) }}
+apiVersion: autoscaling/v2beta2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: {{ template "fullname" . }}
+  namespace: {{ .Release.Namespace }}
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: {{ template "fullname" . }}
+  minReplicas: {{ required "If enabling the HorizontalPodAutoscaler, you must provide .Values.horizontalPodAutoscaler.minReplicas!" .Values.horizontalPodAutoscaler.minReplicas }}
+  maxReplicas: {{ required "If enabling the HorizontalPodAutoscaler, you must provide .Values.horizontalPodAutoscaler.maxReplicas!" .Values.horizontalPodAutoscaler.maxReplicas }}
+  metrics:
+    {{- with .Values.horizontalPodAutoscaler.types }}
+    {{- range $metricType, $metricOptions := . }}
+    - type: {{ if eq $metricType "cpu" "memory" }}Resource{{ else if eq $metricType "custom" }}Object{{ end }}
+      {{- if eq $metricType "cpu" "memory" }}
+      resource:
+        name: {{ $metricType }}
+        target:
+          type: {{ $metricOptions.type }}
+          {{- if eq $metricOptions.type "Utilization" }}
+          averageUtilization: {{ $metricOptions.target }}
+          {{- else if eq $metricOptions.type "averageValue" }}
+          averageValue: {{ $metricOptions.target }}
+          {{- end }}
+      {{- else if eq $metricType "custom" }}
+      object:
+        metric:
+          name: {{ $metricOptions.metricName }}
+        describedObject:
+          apiVersion: apps/v1
+          kind: Deployment
+          name: {{ template "fullname" $ }}
+        target:
+          type: {{ $metricOptions.type }}
+          value: {{ $metricOptions.target }}
+        {{- end }}
+    {{- end }}
+    {{- end }}
+{{- end }}
```

Comment:
@echoboomer Wouldn't this be better to allow the caller to specify whatever they want here?

For example, i would want to make use of https://aws.amazon.com/blogs/compute/scaling-kubernetes-deployments-with-amazon-cloudwatch-metrics/

## Why?
Buildkite already have a way to export metrics into CloudWatch (Lambda, Written in Go). This enables me to use things i already have available to me as metrics to scale up and down my Pods and Nodes.

The current implementation here restrics me to the types `Resource` and `Object` but i'd want to use `External`. Whilst you could extend the `if` statements, it feels like it holds too much logic when the caller can just pass in whatever they want for the `HorizontalPodAutoscaler`

Reply:
I've resolved this in the latest commit. See below.